scripts/git-prehook-whitespace.sh

```diff
@@ -0,0 +1,67 @@
+#!/bin/bash
+#
+# A git hook script to find and fix trailing whitespace
+# in your commits. Bypass it with the --no-verify option
+# to git-commit
+#
+# Copyright Oldman <imoldman.com@gmail.com> and contributors:
+#   https://github.com/imoldman/config
+#
+# Add this to a project's pre-hooks using:
+#  git rev-parse --show-toplevel
+#  ln -s ../../scripts/git-prehook-whitespace.sh .git/hooks/pre-commit
+
+set -euo pipefail
+
+# detect platform
+platform="win"
+uname_result="$(uname)"
+if [ "$uname_result" = "Linux" ]; then
+	platform="linux"
+elif [ "$uname_result" = "Darwin" ]; then
+	platform="mac"
+fi
+
+# change IFS to ignore filename's space in |for|
+IFS="
+"
+# autoremove trailing whitespace
+for line in $(git diff --check --cached | sed '/^[+-]/d') ; do
+	# get file name
+	if [ "$platform" = "mac" ]; then
+		file="$(echo "$line" | sed -E 's/:[0-9]+: .*//')"
+	else
+		file="$(echo "$line" | sed -r 's/:[0-9]+: .*//')"
+	fi
+	# display tips
+	echo -e "auto remove trailing whitespace in \\033[31m$file\\033[0m!"
+	# since $file in working directory isn't always equal to $file in index, so we backup it
+	mv -f "$file" "${file}.save"
+	# discard changes in working directory
+	git checkout -- "$file"
+	# remove trailing whitespace
+	if [ "$platform" = "win" ]; then
+		# in windows, some versions of sed -i can add a read-only attribute
+        # so avoid using -i, and redirect to a new file
+		sed 's/[[:space:]]*$//' "$file" > "${file}.bak"
+		mv -f "${file}.bak" "$file"
+	elif [ "$platform" == "mac" ]; then
+		sed -i "" 's/[[:space:]]*$//' "$file"
+	else
+		sed -i 's/[[:space:]]*$//' "$file"
+	fi
+	git add "$file"
+	# restore the $file
+	sed 's/[[:space:]]*$//' "${file}.save" > "$file"
+	rm "${file}.save"
+done
+
+if [ "x$(git status -s | grep '^[A|D|M]')" = "x" ]; then
+	# empty commit
+	echo
+	echo -e "\\033[31mNO CHANGES ADDED, ABORT COMMIT!\\033[0m"
+	exit 1
+fi
```

Comment:
I'm not sure if this is the bullet-proof way of detecting an empty commit. Git can signify file change by one of the following letters:

[`man git-diff`](https://git-scm.com/docs/git-diff#Documentation/git-diff.txt---diff-filterACDMRTUXB82308203) says:
> (…) Added (A), Copied (C), Deleted (D), Modified (M), Renamed (R), have their type (i.e. regular file, symlink, submodule, ...) changed (T), are Unmerged (U), are Unknown (X), or have had their pairing Broken (B) (…)

Although I don't know which of these are displayed by `git-status`. I would probably use `git-diff HEAD` for detecting an empty commit (not sure yet).

This is not a definitive blocker for review, I just wanted to indicate, that this check might (I can't say it will) fail in some situations, e.g. when commit includes:
- a file that was a symlink but was replaced by a regular file (it could fail on Linux only in such case)
- a submodule added or removed to the project
- a merge commit

Reply:
I knocked out that if block completely; it had nothing to do with stripping trailing white-space. 
Looks like it was put there just as another pre-commit check (ie: block empty commits).  Which I suppose we could also do - but not really a concern either way (plus, empty commits are still valuable for triggering our workflow if something on GitHub's side was broken earlier).